File: langs/scala/runtime/js/src/main/scala/nobleidl/core/JSAdapter.scala
package nobleidl.core

trait JSAdapter[S, J] {
  def toJS(s: S): J
  def fromJS(j: J): S
}

object JSAdapter {
  def identity[A]: JSAdapter[A, A] =
    new JSAdapter[A, A] {
      override def toJS(s: A): A = s
      override def fromJS(j: A): A = j
    }
}

File: langs/java/runtime/src/main/java/dev/argon/nobleidl/runtime/graaljsInterop/JSAdapter.java
package dev.argon.nobleidl.runtime.graaljsInterop;

import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Value;

public interface JSAdapter<T> {
	T fromJS(Context context, JSExecutor executor, Value value);
	Value toJS(Context context, JSExecutor executor, T value);


	static <T> JSAdapter<T> identity() {
		return new IdentityJSAdapter<>();
	}

	static JSAdapter<Value> VALUE_ADAPTER = new JSAdapter<Value>() {
		@Override
		public Value fromJS(Context context, JSExecutor executor, Value value) {
			return value;
		}

		@Override
		public Value toJS(Context context, JSExecutor executor, Value value) {
			return value;
		}
	};

}
